Dockerfile

galexrt opened the pull request:
@timabbott Here you go, a working Dockerfile. Any improvements I could add?

This adds a working Dockerfile and my `create-admin-user.sh` script.

I included the `docker-compose.yml` because the database server needs to be my `galexrt/zulip-postgresql-tsearchextras` (because other postgresql images don't have the tsearch extension installed). You would just need to change the image names (example `zulip/zulip:1.3.10`).

If there is anything that can improve this Pull Request, please let me know.

---

Original commit messages:

```
This adds a working Dockerfile for "production" use
It includes a `docker-compose.yml` which currently uses my image.
The `docker-entrypoint.sh` is about 285 lines of code, but it has inbuilt database backup and restore function.
(I can remove the backup and restore or add more functions see https://github.com/galexrt/docker-zulip)

I also added an README.docker.md, it should give users a quick start on how to use the docker image.

Added Dockerfile for "production" use
Added other docker related files like docker-compose.yml and docker-entrypoint.sh

Added the generate secrets command to the entrypoint

puppet: Move several debugging tool dependencies out of base.pp.

Rename zulip::postgres_appdb to zulip::postgres_appdb_base.

The purpose of this rename is to allow us to move the postgres-related
configuration out of voyager.pp.

puppet: Move tuned postgres configuration out of voyager.pp.

This should make it easier for someone to run just the tuned Zulip
database on one server and the Zulip frontend on another.

puppet: Move prod-static creation from voyager.pp to app_frontend.pp.

Every app frontend will need this directory and this should help
enable more modular puppet rules.

puppet: Move Zulip apt repository to its own manifest.

puppet: Rename app_frontend.pp to app_frontend_base.pp.

This will enable us to move the remaining app-frontend related content
out of voyager.pp.

puppet: Move memcached into its own puppet module.

puppet: Move memcached and rabbitmq include out of app_frontend_base.pp.

puppet: Move default nginx configuration out of voyager.pp.

Document the puppet configuration somewhat in zulip::voyager.

puppet: Make apt repository conditional on the Ubuntu version.

We still will need to address this in the install script as well.

puppet: Use a variable to configure the postgres version.

puppet: Use $postgres_version in postgres template.

This eliminates hardcoding of the postgres version from the Zulip
puppet configuration.

Fix missing puppet dependencies on postgres package.

Added "custom" dockervoyer puppet manifest

Add the files from the repo directly (dont use git clone to get something we already have in our docker build directory)

Added cron.conf for running cron inside the docker container
Useful for later docker related features like auto backup (see https://github.com/Galexrt/docker-zulip/blob/master/entrypoint.sh#L337-L345)

Filled .dockerignore with directories and fiels to ignore/exclude from the image builds

Fixed the linking of directories

Removed git package and don't add "deb-src" list

Fixed the settings path aka the zproject settings folder

Removed the "removement" of "Service" tasks thing puppet

Remove the puppet files after the puppet install

Removed the zproject directory from the $DATA_DIR
The zproject directory causes trouble if it is just linked. Because zulip uses paths like this for example "deployments/../memcached_prefix".

Removed a trailing slash

Wait for database to come online and create database and schema "manually"
(Shouldn't occure because zulip needs the postgres tsearch thingy extensions, thus rendering every other postgresql image/server than the galexrt/zulip-postgresql-tsearchextras:latest)

Added the wait for database code

Fixed the rabbitmq connection problem

Use sed to replace the default memcached location with the env var MEMCACHED_LOCATION

Added the tsearch_extras extension query to the database (re)creation part

Migrate the database everytime we start (for safety)

Tweaks and fixes to the docker-entrypoint.sh

Restructured the entrypoint.sh

Restructured the docker-entrypoint.sh
This restructure adds database backup and restore functionality
Added "Running in Docker" to the README.md

Fixed the help text for the app:help function

Fixed the restore help line (was app:backup is now app:restore)
Added more explanation to the README.docker.md

Small adjustments and tweaks to the entrypoint.sh

We don't need to copy the uploads from the home/zulip

Added restart always to the docker-compose.yml (because there is no wait for container x to become ready)
Added Try counter at the migration

Tweaks and optimization for the entrypoint.sh

Renamed the zulipgeneratesecrets function

This could fix the exit code 0 problem

Fix for the retry at the database migration

Some changes to the sql queries

Small change to the queries
Fixed the if

Added the forgotten nginx.conf to supervisor
(It's a little bit embarassing, that forgot the main part of this container)
```

---

@smarx I signed the CLA.

smarx commented:
*Automated message from Dropbox CLA bot*

@Galexrt, thanks for the pull request! It looks like you haven't yet signed the Dropbox CLA. Please [sign it here](https://opensource.dropbox.com/cla/) and update the thread so we can consider merging your code.

Galexrt commented:
@smarx I signed it (*twice now*)...

smarx commented:
@Galexrt Actually, you've signed it four times. That was the problem... the script used to fail if you had signed more than once (it being unclear which response to use). Now it checks only the most recent signature, so I believe it will pass the next time you try.

timabbott commented:
@Galexrt awesome!  I'm travelling this weekend but will take a look next week.

timabbott commented on Dockerfile:
```diff
@@ -1,12 +1,40 @@
+    chown -R zulip:zulip /home/zulip /var/log/zulip /etc/zulip/settings.py && \
+    apt-get -qq autoremove --purge -y && \
+    apt-get -qq clean && \
+    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
```
It seems like this block of code could be a thin wrapper around `scripts/setup/install`, possible with an option or two added similar to $APT_OPTIONS, right?  Things would be a lot cleaner if we didn't have to maintain a second copy of all this setup code.

timabbott commented on docker-compose.yml:
```diff
@@ -0,0 +1,40 @@
+  volumes:
+    - "/opt/docker/zulip/redis:/var/lib/redis:rw"
+zulip:
+  image: "quay.io/galexrt/docker-zulip:docker"
```
Is an image needed in this version?

timabbott commented on puppet/zulip/files/supervisor/conf.d/nginx.conf:
```diff
@@ -0,0 +1,5 @@
+[program:nginx]
```
Can you add a comment for this one similar to the cron.conf comment?

Galexrt replied:
Comment added.

timabbott replied:
Great, thanks!

timabbott commented on scripts/create-admin-user.sh:
```diff
@@ -0,0 +1,42 @@
+    read -s -p ZULIP_USER_PASS
+done
+
+# Doing everything in python, even I never coded in python #YOLO - Alexander T.
```
Hehe nice first python program :)

timabbott commented on scripts/create-admin-user.sh:
```diff
@@ -0,0 +1,42 @@
+do_change_is_admin(User, True, 'administer')
+User.save()
+quit()
+EOF
```
I don't think you need to write a separate script for this; you should be able to do it with:

`./manage.py create_user <args>` followed by `manage.py knight -f user@example.com`

The only reason that wouldn't do what you want is `manage.py create_user` (implemented in zerver/management/commands/create_user.py) doesn't support passing a password argument; can you instead just add such an option?  I'd happily take that as a separate PR which we can quickly merge; no reason to bundle it in here.

Then this code can be just prompting the user and then running those 2 management commands.

timabbott commented on puppet/zulip/manifests/dockervoyager.pp:
```diff
@@ -0,0 +1,32 @@
+class zulip::dockervoyager {
```
Can you add a brief comment at the top explaining a bit more about what this is?

Galexrt replied:
I added a comment to the file. Is this okay?

timabbott replied:
yep

timabbott commented on README.md:
```diff
@@ -27,6 +27,11 @@ ongoing on adding support for additional platforms. The installation
+===========================
+
+Zulip in Docker. The installation process is documented in [README.docker.md](README.docker.md).
+
```
Can you add an "(experimental)" note here?  And also move this doc to be a subsection of README.prod.md instead of in the main README.md, since that's where we document installing Zulip for production?

timabbott commented on README.docker.md:
```diff
@@ -0,0 +1,74 @@
+Further information on the Zulip project and its features can be found
+at https://www.zulip.org.
+
+Using Zulip with docker
```
Add a "in production" in this heading?

timabbott commented on README.docker.md:
````diff
@@ -0,0 +1,74 @@
+openssl x509 -req -days 3650 -in zulip.csr -signkey zulip.key -out zulip.crt
+```
+
+Congratulations! You have now generated the self-signed SSL certificate, it is valid for 10 years.
````
This section duplicates the SSL part of README.prod.md; can you pull these out into a separate doc (README.ssl.md?) and link to it from both places?

timabbott commented on docker-entrypoint.sh:
```diff
@@ -0,0 +1,217 @@
+    else
+        echo "No database information given. Skipping ..."
+    fi
+}
```
This section duplicates some somewhat tricky code from `scripts/setup/postgres-init-db` -- would it be possible to have this use that script (potentially with an option added to the script)?  I'd prefer to not be maintaining more copies of that database setup code if we can avoid it.

Galexrt replied:
Like in the `scripts/lib/install`, I'll have to add some code there, but that shouldn't be the problem.

timabbott commented:
Cool, thanks for refactoring this, @Galexrt !  This is looking a ton better.  I posted a bunch of comments on how to further simplify the amount of code that this involves maintaining two copies of; I think with those things addressed this'll be fairly clean and reasonable to merge.

This does definitely remind me that we should probably try to get the `tsearch_extras` extension into postgres upstream, since maintaining a postgres extension module is kinda a pain.

Galexrt commented:
@timabbott I had to chang the `scripts/lib/install` a bit, so the Docker installation is possible without wrapping the code around (see your comment https://github.com/zulip/zulip/pull/450#discussion_r51826477).

Galexrt commented:
Adding the `tsearch_extras` to postgres upstream would be good, because it's a pain to maintain the special needed image [Galexrt/docker-zulip-postgresql-tsearchextras](https://github.com/Galexrt/docker-zulip-postgresql-tsearchextras). The image [Galexrt/docker-zulip-postgresql-tsearchextras](https://github.com/Galexrt/docker-zulip-postgresql-tsearchextras) is needed because of the `tsearch_extras`, but it's not that much trouble.

Galexrt commented:
@timabbott What package do I need to install to generate the `prod-static` files?
(I currently get the `prod-static` files through, downloading latest zulip version and extracting `prod-static` files)

timabbott commented:
They're generated by `tools/update-prod-static`.

That said, it seems like by default the Docker-in-production setup should pull from a release tarball, not Git, since that's likely to be more stable for a production environment.  It'd be great to have an option for it to deploy from Git though.

Galexrt commented:
@timabbott I added the `Dockerfile-prod`. The `Dockerfile-prod` uses the latest release tarball as you want, but building the Docker image will first work when the changes I made to the scripts are in the release tarball.

**Dockerfile Overview:**
* `Dockerfile` uses the files from the directory ("branch" building)
* `Dockerfile-prod` uses latest release tarball
* `Dockerfile-dev` is the old `Dockerfile`

Are the multiple Dockerfiles for different "builds" okay?

Galexrt commented:
@timabbott I get the following errors when running the `update-prod-static` tool.
```
zulip_1     | npm WARN EPACKAGEJSON zulip@0.0.0 No description
zulip_1     | npm ERR! Linux 4.1.16-1-MANJARO
zulip_1     | npm ERR! argv "/usr/bin/nodejs" "/usr/bin/npm" "install"
zulip_1     | npm ERR! node v5.5.0
zulip_1     | npm ERR! npm  v3.3.12
zulip_1     | npm ERR! code ELIFECYCLE
zulip_1     | 
zulip_1     | npm ERR! contextify@0.1.15 install: `node-gyp rebuild`
zulip_1     | npm ERR! Exit status 1
zulip_1     | npm ERR! 
zulip_1     | npm ERR! Failed at the contextify@0.1.15 install script 'node-gyp rebuild'.
zulip_1     | npm ERR! Make sure you have the latest version of node.js and npm installed.
zulip_1     | npm ERR! If you do, this is most likely a problem with the contextify package,
zulip_1     | npm ERR! not with npm itself.
zulip_1     | npm ERR! Tell the author that this fails on your system:
zulip_1     | npm ERR!     node-gyp rebuild
zulip_1     | npm ERR! You can get their info via:
zulip_1     | npm ERR!     npm owner ls contextify
zulip_1     | npm ERR! There is likely additional logging output above.
zulip_1     | 
zulip_1     | npm ERR! Please include the following file with any support request:
zulip_1     | npm ERR!     /home/zulip/deployments/2016-02-05-14-43-28/npm-debug.log
zulip_1     | Traceback (most recent call last):
zulip_1     |   File "/home/zulip/deployments/current/tools/update-prod-static", line 29, in <module>
zulip_1     |     subprocess.check_call(['npm', 'install'], stdout=fp, stderr=fp);
zulip_1     |   File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
zulip_1     |     raise CalledProcessError(retcode, cmd)
zulip_1     | subprocess.CalledProcessError: Command '['npm', 'install']' returned non-zero exit status 1
```

timabbott commented:
I think the overall issue here is that there are some dependencies needed to build the static assets that we don't install in the production puppet files (since they aren't needed there).  These include:
* `closure-compiler` (currently doesn't show up in puppet at all, which I should probably fix)
* nodejs
* npm
* yui-compressor
* libfreetype6-dev
* (possibly some more of the libraries from the list in provision.py?  Those are the ones I'm confident of)
I think probably the right fix for this is to add a new puppet module, e.g. `puppet/zulip/manifests/static_build.pp` that installs the dependencies needed to run the static asset pipeline.  

That error in particular seems like it could be e.g. a permissions problem (that's what I see when i google ELIFECYCLE); I'd check the npm-debug.log to try to figure out what's up...

Galexrt commented:
@timabbott I have removed the `npm build` from the `Dockerfile`'s and switched to using the Zulip tarball.

Galexrt commented:
I just tested the image, with run `docker-compose up`, configured all configs it and restarted the containers. Et voilà a working Zulip instance. :)

The only thing that wasn't properly working, are the emoticons. They appear in the auto completion but not when the message is sent, but I'll look into this now.

Galexrt commented:
@timabbott Any news about merging this?

Galexrt commented:
The `Dockerfile-prod` will only work when there is a new release with the modified scripts and the added `dockervoyager.pp` in it .

gvt commented:
+1 for these changes. would like to see this released so that docker-zulip can use it and fix the error.

timabbott commented on scripts/lib/install:
```diff
@@ -13,49 +22,60 @@ EOF
+    # We generate the static files later in the entrypoint.sh we don't have the
+    # static files right now, so just create the folder to fix the error with "cp ..."
+    mkdir -p "$deploy_path"/prod-static/serve
+fi
```
Is this still needed with the new model of using the static assets from a release tarball?

Galexrt replied:
You are right, this sliped through the change back to using the release tarball. Changed now.

timabbott commented on scripts/setup/postgres-init-db:
```diff
@@ -2,32 +2,74 @@
+# Stop supervisord only if the socket exists, prevents problems with docker
+if [ -e "/var/run/supervisor.sock" ]; then
+    supervisorctl stop all
+fi
```
Can you explain the problem with docker this causes?  Is it just supervisorctl throwing an error if it wasn't already running?

Galexrt replied:
While building the image, supervisor won't be started. That causes `supervisorctl stop all` to return != 0 and exit the script with an error, this prevent this from happening.

timabbott commented on scripts/setup/postgres-init-db:
```diff
@@ -2,32 +2,74 @@
+    # Drop any open connections to any old database.  Hackishly call using
+    # source because postgres user can't read /root/zulip/scripts/setup.
+    source "$(dirname "$0")/terminate-psql-sessions" postgres zulip zulip_base
+fi
```
What happens in the docker case in docker -- does this just do nothing there, or what?

Galexrt replied:
> While building the image, supervisor won't be started. That causes supervisorctl stop all to return != 0 and exit the script with an error = build failure, this prevent this from happening.
See my other comment on this file.

timabbott replied:
OK on the supervisor front I think it's reasonable to check for the socket -- that's a correct check for whether supervisord is running.

For the `terminate_psql_sessions` part, I'm not sure why that would be a problem in the docker setting -- the next lines in this script only work if the script can't access to the database anyway...

Galexrt replied:
I think that the `terminate_psql_sessions` fails because it tries to connect to the localhost "only". If the script fails, the "setup" of the image fails and won't start. 

I'm going to check if I can remove the `if [ "$DEPLOYMENT_TYPE" != "dockervoyager" ]; then` here.

timabbott commented:
Sorry for the slow review here, folks!  This got caught during the window just before a 2-week vacation, and when I got back I've been working through all the Google Summer of Code traffic on the project.  I posted several more comments on this; I think @Galexrt also needs to investigate the emoji not working properly.

Galexrt commented:
@timabbott I'm looking into the emoji problem.

timabbott commented:
We should discuss entrypoint.sh but I'd like to focus effort for now on getting the basic functionality merged first, so we can minimize merge conflict risk, etc.

Galexrt commented:
@timabbott Okay, should I open a fresh pull request with the changes to the script only and add the Docker stuff in this PR?

timabbott commented:
Nah just keep updating this PR and I'll start merging pieces of it as they're ready...

Galexrt commented:
@timabbott I have now fixed the emoji problem, pushed my `entrypoint.sh` in and rebased on the latest changes. Is there anything I can still improve?

![2016-03-26-124221_1517x1034_scrot](https://cloud.githubusercontent.com/assets/3718398/14059845/39c8ae4c-f350-11e5-93b0-1433506c6175.png)
The picture is from my `galexrt/docker-zulip`, that is now fully included in this PR (it is working because I manually add the files from this commit to it).

The image currently "can't" be built, because my changes in the scripts, aren't yet in a release tarball. When my changes to the scripts and puppet are in a new release tarball, it'll build without problems (I have tested it with my modified `galexrt/docker-zulip` image).

This image is now **100% compatible** to my `galexrt/docker-zulip` image. So all users of my image can switch (but no necessary to switch) to the official Zulip image with 0 effort.

timabbott commented:
Awesome!  OOC what was the emoji bug?

I expect to have time to review this in detail in the next week; I plan to start by fully reviewing and merging the `install` and `postgres` script changes since those are the most likely to merge conflict with other things folks are doing.

Galexrt commented:
@timabbott The problem was I created the wrong symlinks after the `install` script. After I removed my code creating links to the prod-static folder it worked perfectly.

timabbott commented:
Yeah the build failure is an issue with how we were getting phantomjs packages; if you rebase onto master the issue is fixed.

Galexrt commented:
I screwed up the rebaseing badly, but I fixed it now. I'm just waiting for Travis to succeed.

galexrt commented:
Rebased again.

galexrt commented:
Rebased on latest Zulip changes.

@timabbott Ping.

timabbott commented on scripts/lib/install:
```diff
@@ -1,61 +1,81 @@
+# Call the default type "voyager", for docker it's "dockervoyager"
+DEPLOYMENT_TYPE="${DEPLOYMENT_TYPE:-voyager}"
+# Can be for example "dockervoyager" or voyager
+VOYAGER_CLASS="${VOYAGER_CLASS:-voyager}"
```
I think we should include the `zulip::` in here, and rename it to something like "PUPPET_CLASSES", because this is actually a comma-separated list, not a single value.  E.g. you might want "zulip::app_frontend, zulip::memcached"

galexrt replied:
@timabbott I renamed the variable, but I think it is better to just "call" the main classes (`zulip::voyager` or `zulip::dockervoyager`) here.

timabbott commented:
Yeah, sorry for the delay reviewing this!  If you've been following the project, you've probably seen there's been a ton of new pull requests recently, and I've been prioritizing the quick ones :/

I won't get to do more on this this week, but next week should be a lot better.

timabbott commented:
OK I've finally found some time to look at this some more.  I'm getting ready to merge the changes to `install` and `postgres-init-db` (have been splitting them into individual readable commits here: https://github.com/timabbott/zulip/commits/docker).

One thing I noticed that I find very confusing is this comment in `postgres-init-db`:
+# In Docker we have to run this script everytime we start the container, so don't drop.

Why do you need to run `postgres-init-db` every time you start the container?  It seems like ideally this would only ever need to be run once.

timabbott commented on scripts/setup/terminate-psql-sessions:
```diff
@@ -1,6 +1,5 @@
 #!/usr/bin/env bash
-set -e
-set -x
+set -ex
```
We don't need this hunk, right?  Can you drop it and the similar ones?

timabbott commented on zerver/lib/queue.py:
```diff
@@ -260,4 +260,3 @@ def queue_json_publish(queue_name, event, processor):
             get_queue_client().json_publish(queue_name, event)
         else:
             processor(event)
-
```
Can you drop this whitespace change?

timabbott commented on Dockerfile:
```diff
@@ -1,12 +1,26 @@
+EXPOSE 80 443
+
+ENTRYPOINT ["/sbin/entrypoint.sh"]
+CMD ["app:run"]
```
Renaming so the production docker file has the "Dockerfile" name is potentially fine, but it might be better to have "Dockerfile-prod" and "Dockerfile.dev" for clarity.  Regardless, I think the README.dev.md instructions for using the old development Dockerfile need to be updated to specify the path to the Dockerfile....

timabbott commented:
I merged the `install` and `postgres-init-db` changes that don't check DEPLOYMENT_TYPE, since I think those are all good.  Thanks @galexrt!

Still waiting to understand what's going on with how `postgres-init-db` works in this docker configuration before I try to merge the rest of the changes to those scripts.

galexrt commented:
@timabbott Sorry for the late response.
In the container the `postgres-init-db` is currently only run in the `install` script, that's why there is the exclusion to not run the `postgres-init-db` script when `DEPLOYMENT_TYPE` is `dockervoyager`.

The reason I added those changes to the `postgres-init-db` script is, so that I can remove the psql queries from the `docker-entrypoint.sh` and let the `postgres-init-db` script do the database setup job. I have just now seen that I forgot to replace the queries in the `docker-entrpoint.sh`. I'm going to push this change with the next rebased commit.

I'm going to add the improvements and fixes you pointed out in the comments, in that commit too.

galexrt commented:
@timabbott I have now pushed the changes from the comments.
I'm currently rebasing on the master branch.

galexrt commented:
@timabbott I just pushed my rebased `docker` branch. I had readded my changes to the `postgres-init-db`, but now after checking in the `docker-entrypoint.sh` I have now reverted the `postgres-init-db` back to the state in your `docker` branch.

timabbott commented on README.dev.md:
````diff
@@ -380,7 +380,7 @@ Then you should create the Docker image based on Ubuntu Linux, first
 
 ```
-docker build -t user/zulipdev .
+docker build -f Dockerfile-dev -t user/zulipdev .
````
Does the user need to pass `-f Dockerfile-dev` when using `docker run` as well?

galexrt replied:
The `-f Dockerfile-dev` is only me needed for the `docker build` and not for the `docker run` command.

timabbott commented:
I'm really really sorry @galexrt, but I haven't had a chance to work on this more before a vacation for 2 weeks I'm leaving for tomorrow.  This is the pull request that I've been most behind on finding time to review, so I'll try hard to find time to look in detail when I get back.

galexrt commented:
No problem. I'm going to keep the PR rebased as best as I can. Have a good vacation!

galexrt commented:
I don't see why the backend build is failing because this comment doesn't change any backend stuff..

timabbott commented:
If you click on "Details" you can actually read the error messages.  The `backend` suite also runs the Zulip linters; the failure is:
```
+./tools/lint-all
No newline at the end of file.  Fix with `sed -i '$a\' docker-entrypoint.sh`
The command "./tools/travis/$TEST_SUITE" exited with 1.
```

galexrt commented:
@timabbott Thanks for the heads up!
I just added a new line to the `docker-entrypoint.sh`.

timabbott commented:
@galexrt I did a bit of related work that I wanted to point you to in 02191491830a34447da379fba248dd0d59fac6f1 (currently on my branch but should be merged soon); see https://github.com/zulip/zulip/issues/1235 for how to use it.  This should help with eventually making it possible to have a set of docker files to run each Zulip service in its own instance.

galexrt commented:
@timabbott That's awesome!
Could you please mention me when the changes are merged so that I can modify the `Dockerfile` to match your changes.
The `dockervoyager.pp` is still needed with your changes in mind. Is that okay?

timabbott commented:
Yep, can do and totally agreed that we'll still need dockervoyager.pp.

timabbott commented on .dockerignore:
```diff
@@ -0,0 +1,11 @@
+THIRDPARTY
+Vagrantfile
+frontend_tests/
+docker-compose*.yml
```
Would it work to just make this a symlink to `.gitignore`?

timabbott replied:
If not, can you add a comment at the top of this explaining how one should decide which things need to go in here?

galexrt replied:
Comment added.

rht replied:
note: the file can be autogenerated with https://github.com/jamestalmage/dockerignore, which reduces human err when doing the maintenance.

timabbott commented on README.docker.md:
```diff
@@ -0,0 +1,56 @@
+Zulip
```
Can you move this to somewhere under `docs/` as part of our ReadTheDocs tree?

timabbott commented on README.dev.md:
````diff
@@ -41,7 +41,77 @@ See [this guide](https://zulip.readthedocs.io/en/latest/install-generic-unix-dev
+```
+
+You can modify the source code in your development machine and review
+the results in your browser.
````
I think these changes are a mis-resolved merge conflict with our moving this content to `docs/install-docker-dev.md`.  Can you address that?

timabbott commented:
@galexrt I'd like to merge some more pieces of this.  Here's a proposed way to split this up into mergable commits:
* A commit that just renames the existing Dockerfile to Dockerfile.dev and updates the existing development docs accordingly
* A commit that just adds the puppet stuff (dockervoyager.pp and its dependencies).
* A commit adding the .dockerignore file
* A commit adding Dockerfile, plus the documentation on how to use this.

timabbott commented on Dockerfile-prod:
```diff
@@ -0,0 +1,27 @@
+FROM quay.io/sameersbn/ubuntu:latest
+MAINTAINER Alexander Trost <galexrt@googlemail.com>
+
+ENV ZULIP_VERSION="1.3.10" DATA_DIR="/data"
```
For `ZULIP_VERSION`, would it make sense to use `zulip-server-latest.tar.gz` by default but have a way for the user to pass in a different version as an override?  Then we don't need to update this when we do Zulip releases.

galexrt replied:
I would just change the `ZULIP_VERSION` to `latest` so it always downloads and installs the latest version or the user specified.

timabbott commented on Dockerfile-prod:
```diff
@@ -0,0 +1,27 @@
+    tar xfz /tmp/zulip-server.tar.gz -C /root/zulip --strip-components=1 && \
+    rm -rf /tmp/zulip-server.tar.gz && \
+    export VOYAGER_CLASS="dockervoyager" DEPLOYMENT_TYPE="dockervoyager" \
+        ADDITIONAL_PACKAGES="python-dev python-six python-pbs" has_nginx="0" has_appserver="0" && \
```
`python-pbs` is no longer used in Zulip, so you can probably drop it here.

timabbott commented on Dockerfile-prod:
```diff
@@ -0,0 +1,27 @@
+    export VOYAGER_CLASS="dockervoyager" DEPLOYMENT_TYPE="dockervoyager" \
+        ADDITIONAL_PACKAGES="python-dev python-six python-pbs" has_nginx="0" has_appserver="0" && \
+    /root/zulip/scripts/setup/install && \
+    chown zulip:zulip /opt/createZulipAdmin.sh && \
```
where does createzulipadmin.sh come from?

galexrt replied:
It is a script that creates a zulip user with admin perms. The script is from my `galexrt/docker-zulip` repo.
Should I include it so Docker users have a simple way of creating an user?

timabbott commented on puppet/zulip/manifests/dockervoyager.pp:
```diff
@@ -0,0 +1,36 @@
+    }
+  }
+  safepackage { $appdb_packages: ensure => "installed" }
+  include zulip::supervisor
```
is there a reason `zulip::supervisor` is down here?

galexrt replied:
Not really. I'm moving it up.

timabbott commented:
Posted a few comments on various things.  The big issue here is that I'm still not sold on how much duplicated code `docker-entrypoint.sh` has -- e.g. why can't `bootstrapRabbitMQ` just shell out to  `scripts/setup/configure-rabbitmq` (perhaps with some new options added to support making the usernames/etc. variables, and similarly for the initial database setup)?  At present, it seems like there'd just going to be maintenance work involved with updating that script every time we refactor Zulip's setup process, which could be avoided.  But perhaps I'm missing something here.

galexrt commented:
@timabbott I'm going to look into de-duplicate the code from the entrypoint by adding some changes to the configuration scripts of Zulip in a separate PR.

galexrt commented:
I finally had some time to rebase and remove the rabbitmq and initialize-database duplicated code from the entrypoint.

galexrt commented:
I rebased the commit. There is one thing that still causes problems.
The supervisor service fails to start on certain Docker build environments (like Docker Hub), because the supervisor service can't start. Is there a way to not start the supervisor service in puppet when running in a container?

galexrt commented:
Fixed the docs.

***

@timabbott Any idea how I can stop Puppet from restarting the Supervisor service when in running in Docker build?

timabbott commented:
Sorry for the slow reply; can you explain the context/problem a bit more @galexrt?  Is this puppet running inside a container that is running the main app (and thus has supervisord included) or one of the other containers?

galexrt commented:
The problem is that when Puppet tries to start supervisor service it fails. Resulting in the whole deployment to fail with a non zero exit code.
I may have found a way to fix this.

galexrt commented:
I have added a variable named `$ignoreSupervisorService` to "create" the supervisor service but with `/bin/true` as the start/stop handlers.
This fixes the issues with supervisor in the Puppet deployment not starting when building the Docker image.
Is this okay?

galexrt commented:
@timabbott Puppet lint doesn't show me any errors, but Travis `lint-all` fails with:
```
Error: Could not parse for environment production: Syntax error at ';'; expected '}' at /home/travis/build/zulip/zulip/puppet/zulip/manifests/dockervoyager.pp:21
```
Is that a syntax error or not?

galexrt commented:
@timabbott Could you tell me about the error I got in the Travis build?

timabbott commented:
Travis CI just changed their build environments overnight, it's not your issue.  Working on it.

jezdez commented:
@timabbott @galexrt Hey there, I've been following along the effort to make Zulip play better with Docker and wondered if you considered using one of the official [Postgres images](https://hub.docker.com/_/postgres/) instead of the fork of a custom one that @galexrt has been using for a while now.

I'm asking since I'm interested in the long term maintenance of a Docker based deployment of Zulip where a wholly custom Postgres Docker image could be a risk and lead to the increase of TCO. I realize that this PR has been coming for a long time, apologies for this late feedback, @galexrt. 🙇

Just out of curiosity I [checked the diff](https://github.com/sameersbn/docker-postgresql/compare/master...galexrt:master) between @galexrt's Postgres fork and the [main repo](https://github.com/sameersbn/docker-postgresql) of the custom Postgres repo and found it not too complicated to adopt to the official Postgres docker image. Luckily the image is able to be extended fairly easy (see ["How to extend this image"](https://github.com/docker-library/docs/tree/master/postgres#how-to-extend-this-image)).

So I think it boils down to:

- installing the `tsearch_extras` package
- hooking up the dictionary and stop words
- add a shell script that adds the zulip schema and enables unaccent and tsearch_extras extions

The result is https://github.com/jezdez/postgres-zulip-tsearchextras if you want to take a look at. If this is hooked up with CircleCI or Travis CI to push to Docker Hub whenever @timabbott does a release of tsearch_extras I think this should be the most stable solution going forward. What do you think?

galexrt commented:
@jezdez It sound much better than the my current image.
I'll update the `docker-compose.yml` in this PR (and my galexrt/docker-zulip repo).

galexrt commented:
@timabbott Now that I fixed the text search issue and database issue.
Could you please take another look at this?

timabbott commented on docs/install-docker-dev.md:
```diff
@@ -1,3 +1,75 @@
 # Using Docker (experimental)
 
 Moved to [Advanced setup (Docker)](dev-setup-non-vagrant.html#using-docker-experimental).
+=======
```
Is this from a merge conflict?

timabbott commented on scripts/setup/configure-rabbitmq:
```diff
@@ -5,11 +5,12 @@
+sudo rabbitmqctl -n "$RABBITMQ_HOST" delete_user guest || true
+sudo rabbitmqctl -n "$RABBITMQ_HOST" add_user "$RABBITMQ_USERNAME" "$RABBITMQ_PASSWORD"
+sudo rabbitmqctl -n "$RABBITMQ_HOST" set_user_tags "$RABBITMQ_USERNAME" administrator
+sudo rabbitmqctl -n "$RABBITMQ_HOST" set_permissions -p / "$RABBITMQ_USERNAME" '.*' '.*' '.*'
```
can you extract this as its own commit?  I sorta thought we'd done this, but my memory may be off.

galexrt replied:
Done. Here you go https://github.com/zulip/zulip/pull/4546

timabbott commented on puppet/zulip/manifests/supervisor.pp:
```diff
@@ -3,31 +3,43 @@
+      status     => "/bin/true",
+      hasrestart => true,
+      restart => "/bin/true"
+    }
```
Can you add a quick comment explaining this section for folks who might not know what's going on?

galexrt replied:
Added a comment.

timabbott commented on puppet/zulip/manifests/dockervoyager.pp:
```diff
@@ -0,0 +1,41 @@
+  include zulip::app_frontend
+
+  $appdb_packages = [# Needed to run process_fts_updates
+    "python-psycopg2",
```
I think this may no longer be needed?  I think `process_fts_updates` knows how to find the Zulip virtualenv.

timabbott commented on .dockerignore:
```diff
@@ -0,0 +1,44 @@
+static/js/bundle.js
+static/generated/emoji
+static/generated/github-contributors.json
+static/locale/language_options.json
```
Several of these files are needed in production, but can be generated; can you add a comment explaining these?

It might make sense to use the approach we use in `tools/build-release-tarball`, where we start by basically using `git archive` to create a fresh clone-type tarball/directory from the current git commit.  That would eliminate the need for this file to duplicate `.gitignore`.

timabbott commented on docs/install-docker-production.md:
```diff
@@ -0,0 +1,54 @@
+The default data volume location of the Zulip container on the host system is `/opt/docker/zulip/zulip`.
+The default data volume location of the PostgreSQL server on the host system is `/opt/docker/zulip/postgresql`.
+The default data volume location of the Redis server on the host system is `/opt/docker/zulip/redis`.
+To change it modify the path(s) in the `docker-compose.yml`.
```
Can you expand this with some more notes about how to avoid losing your data when using Docker in production?

galexrt replied:
Added a comment.

timabbott commented:
Cool,  @galexrt what are your thougths on my question about whether there's a way to avoid the duplication of the install script and settings file in `docker-entrypoint.sh`?

galexrt commented:
@timabbott As a note: I still have this on my radar, but I currently have some exams coming up. I should have time to implement the new feedback in about 2 weeks from now.

galexrt commented on docs/install-docker-dev.md:
```diff
@@ -1,3 +1,75 @@
 # Using Docker (experimental)
```
I would leave the file in as I'm going to expand it later on when the Docker dev environment has got its revamp/improvements.

rht commented on Dockerfile:
```diff
@@ -1,19 +1,27 @@
-RUN echo 'export LC_ALL="en_US.UTF-8" LANG="en_US.UTF-8" LANGUAGE="en_US.UTF-8"' >> ~zulip/.bashrc
-
-WORKDIR /srv/zulip
+FROM quay.io/sameersbn/ubuntu:latest
```
Would it be better to pin the version, e.g. at `14.04.20170608`, for a reproducible build?

galexrt replied:
Yes that would be better.

rht commented on Dockerfile:
```diff
@@ -1,19 +1,27 @@
+    export VOYAGER_CLASS="dockervoyager" DEPLOYMENT_TYPE="dockervoyager" \
+        ADDITIONAL_PACKAGES="python-dev python-six" has_nginx="0" has_appserver="0" && \
+    /root/zulip/scripts/setup/install && \
+    chown zulip:zulip /opt/createZulipAdmin.sh && \
```
?

galexrt replied:
What line do you mean in specific?

timabbott commented:
Closing, since a lot of the pieces of this have already been merged through other PRs.  

@galexrt I'd definitely welcome a cleaned-up version of this being resubmitted :)
